# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
#группируем данные по кол-ву детей с выводом количества заемщиков и доли должников среди них 
#в виде сводной таблицы с помощью groupby: долю можно вывести сразу через mean или вывести
#количество должников и вывести результаты расчетов в отдельной таблице - см. закомментированный вариант
#к сводной таблице для наглядности добавим столбец со значениями доли каждой группы заемщиков от общего числа

data_groupedby_children = data.groupby('children').agg({'debt': ['count', 'mean']})

#data_groupedby_children = data.groupby('children').agg({'debt': ['count', 'sum']})

#debt_children_ratio = data_groupedby_children['debt']['sum'] / data_groupedby_children['debt']['count']

data_groupedby_children['children_part'] = data_groupedby_children['debt']['count'] / data_groupedby_children['debt']['count'].sum()

data_groupedby_children


debt           children_part
          count      mean              
children                               
0         14091  0.075438      0.660588
1          4808  0.092346      0.225400
2          2052  0.094542      0.096198
3           330  0.081818      0.015470
4            41  0.097561      0.001922
5             9  0.000000      0.000422

**Вывод:** в целом можно сказать, что процент должников среди бездетных заемщиков (7.5%) ниже, чем среди тех, у кого 1-4 ребенка (медианное значение около 9.3%), при этом доля должников среди тех, у кого трое детей немного снижается (до 8.2%) относительно тех, у кого 1, 2 и 4 ребенка (медиана 9.4%). Что касается тех, у кого 5 детей, то доля должников среди них нулевая. Однако само количество заемщиков значительно снижается с прибавлением каждого следующего ребенка, так бездетные составляют 66.1% заемщиков банка, люди, имеющие 1 ребенка уже всего 22.5%, 2х - 9.6%, 3х - 1.5%, 4х - 0.2%, а 5х вообще менее десятой доли % - это всего 9 человек против тысяч среди тех, у кого не более 2 детей и то, что среди них не встретилось должников не гарантирует, что данное соотношение будет релевантно для большей выборки. 9 человек - слишком маленькая группа, чтобы отличить закономерность от шума, поэтому данными по этой группе можно пренебречь. Тем не менее, тот факт, что доля должников среди имеющих 1-4 детей выше почти на 2%, чем среди бездетных даже при значительно меньшем количестве группы, позволяет говорить о более предпочтительном месте в кредитном рейтинге для бездетных, относительно заемщиков с детьми.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
data_groupedby_family_status = data.groupby('family_status').agg({'debt': ['count', 'mean']})

data_groupedby_family_status['family_status_part'] = data_groupedby_family_status['debt']['count'] / data_groupedby_family_status['debt']['count'].sum()

data_groupedby_family_status 


debt           family_status_part
                       count      mean                   
family_status                                            
Не женат / не замужем   2796  0.097639           0.131077
в разводе               1189  0.070648           0.055740
вдовец / вдова           951  0.066246           0.044583
гражданский брак        4134  0.093130           0.193802
женат / замужем        12261  0.075606           0.574797

**Вывод:** Самые низкие доли должников наблюдаются среди групп: "вдовец/вдова", "в разводе" и "женат/замужем" (6.6%, 7.1% и 7.5% соответственно). При этом люди, состоящие в браке составляют 57.4% заемщиков, а бывшие в браке наоборот наименьшие доли суммарно (в соотношении почти 50/50) дающие почти 10% от общего количества. Ощутимо выше доля должников среди заемщиков, живущих в гражданском браке - 9.3% при доле в общей массе в 19.3% и, наконец, наибольший процент неплательщиков (9.7%) среди не состоящих и не состоявших ни в каком браке (при их доле в общей массе в 13.1%). Возможно, прослеживается связь между долей ответственности, которую людям уже приходилось брать на себя в связи с личными отношениями и степенью ответственности, с которой они подходили к кредитным вопросам. Так люди, состоящие или прошедшие через брак более ответственны к своим кредитным обязательствам, а не женатые официально менее ответственны. При этом больше всего кредитами пользуются люди, состоящие в официальном браке, на втором месте те, кто состоит в гражданском браке (почти в 3 раза меньше) и уже в 3 учередь не состоящие и не состоявшие ни в каком браке (более, чем в 4 раза меньше). Похоже, что наличие второй половины создает ощутимо больше спроса на кредиты и чем более узаконены узы, тем выше спрос.     

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
#действуем аналогично предыдущему заданию

data_groupedby_total_income = data.groupby('total_income_category').agg({'debt': ['count', 'mean']})

#data_groupedby_total_income = data.groupby('total_income_category').agg({'debt': ['count', 'sum']})

#debt_total_income_ratio = data_groupedby_total_income['debt']['sum'] / data_groupedby_total_income['debt']['count']

data_groupedby_total_income['total_income_part'] = data_groupedby_total_income['debt']['count'] / data_groupedby_total_income['debt']['count'].sum()

data_groupedby_total_income 


debt           total_income_part
                       count      mean                  
total_income_category                                   
A                         25  0.080000          0.001172
B                       5014  0.070602          0.235057
C                      15921  0.084982          0.746379
D                        349  0.060172          0.016361
E                         22  0.090909          0.001031

**Вывод:** почти 3/4 заемщиков сосредоточены в диапазоне доходов свыше 50000 и до 200000 руб. включительно. Еще почти четверть за диапазоном более 200000 и до 1 млн. руб включительно. Все остальные группы составляют незначительные доли, а количество людей в них не позволит отличить шум от закономерности. Поэтому для определения зависимости корректно будет использовать только группы C и B. При том, что заемщиков с уровнем доходов свыше 200 тыс. руб. в 3 раза меньше (что логично, так как и людей с такими доходами в целом меньше), наблюдается снижение доли должников (т.е. растет надежность заемщиков) по мере роста уровня доходов. Помимо того, что с ростом доходов растет платежеспособность людей, возможно, также растет и уровень финансовой грамотности, умения трезво оценивать свои силы и ответственности / важности сохранения репутации. 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
data_groupedby_purpose_category = data.groupby('purpose_category').agg({'debt': ['count', 'mean']})

data_groupedby_purpose_category['purpose_category_part'] = data_groupedby_purpose_category['debt']['count'] / data_groupedby_purpose_category['debt']['count'].sum()

data_groupedby_purpose_category 


debt           purpose_category_part
                          count      mean                      
purpose_category                                               
операции с автомобилем     4279  0.093480              0.200600
операции с недвижимостью  10751  0.072551              0.504008
получение образования      3988  0.092528              0.186958
проведение свадьбы         2313  0.079118              0.108434

**Вывод:** ожидаемо наиболее популярны кредиты на недвижимость - 50% от общей массы, при этом и надежност таких заемщиков наиболее высока - 7.3% должников - видимо сказывается уровень серьезности и взвешенности решений, связанных с вопросами "крыши над головой". Немного хуже рейтинг у заемщиков для целей проведения свадьбы - 7.9% должников (при доле в 10.8% от общей массы кредитов). Разница не столь велика - видимо сказывается серьезность принятых на себя обязательств по созданию новой "ячейки общества", возможная помощь родителей и нежелание омрачать прекрасный повод разбирательствами с банком. На 3м месте по отсутствию должников заемщики для целей получения образования - уже 9.3% (при доле в 18.6% от общей массы кредитов). Возможно сказывается (теоретически) более низкий возраст, жизненный опыт и уровень дохода целевых потребителей подобных кредитов, однако и помощь родителей может выручать с другой стороны. И с небольшой разницей в худшую сторону 4 место занимают заемщики на операции с автомобилем - 9.3% должников (при доле в 20% от общей массы). Несмотря на поговорку, при нынешних ценах на машины - это скорее роскошь, а не средство, особенно в сравнении с другими 3 группами. И видимо людям реже удается трезво оценить свои финансовые возможности в погоне за комфортом передвижения или неким статусом, игрушкой для взрослых (в некоторых случаях).  

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* пропуски в равном количестве присутствовали в столбцах с трудовым стажем и доходом заемщиков. Анализ данных показал, что пропуски в этих столбцах допущены в одних и тех же строках и нет явной зависимости наличия пропусков от типа занятости, возраста или пола заемщика (см. комментарии к задаче 4). Можно предположить, что пропуски носят либо технический характер (так как не были выгружены данные по одним и тем же строкам), либо конкретные заемщики по каким-то причинам не предоставляли данную информацию (но в таком случае не совсем понятно, каким образом они смогли подтвердить свою платежеспособность для одобрения выдачи кредита банком, поэтому данный вариант менее вероятен).  

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* в случае применения среднего арифметического крайние значения, сильно отличающиеся от основной группы могут сильно стянуть в сврою сторону среднее значение. Медиана как раз выбирает среднее, исходя из большинства имеющихся значений и подставив такое число в общий ряд вместо пропусков мы меньше рискуем получить отклонение от основной группы (создать больше шума).

### Шаг 4: общий вывод.

**Общий вывод:** Определенная зависимость шансов на возврат кредита от количества детей, семейного положения, дохода и целей кредита наблюдается. Бездетные заемщики (составляют 66.1% от общей массы), по имеющимся данным, выглядят предпочтительнее тех, у кого есть дети. Люди, бывшие в браке или в нем состоящие (в сумме составляющие порядка 66%) предпочтительнее свободных и живущих в гражданском браке. Кредиты на недвижимость и свадьбы возвращаются стабильнее (и составляют порядка 61% от общего числа), чем на образование и автомобили. 
Чем больше уровень дохода, тем чаще заемщики возвращают кредиты. Почти 75% заемщиков сосредоточены в диапазоне доходов свыше 50000 и до 200000 руб. включительно. Еще почти 25% за диапазоном более 200000 и до 1 млн. руб включительно.